flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptivebatch/AdaptiveBatchScheduler.java

```diff
@@ -149,13 +159,45 @@ protected void startSchedulingInternal() {
         super.startSchedulingInternal();
     }
 
+    public boolean updateTaskExecutionState(TaskExecutionState taskExecutionState) {
+        updateResultPartitionBytesMetrics(taskExecutionState.getIOMetrics());
```

Comment:
1, 2：Fixed.
3：This is indeed a problem. We can't clear the metric of a partition when it is reset, because for `ALL_TO_ALL` results, we stores aggregated value. Given that, I think we should only record the metric on the first finish, and let it as the final value. @zhuzhurk  WDYT ?

Reply:
Looks to me it's not right because the data produced by an execution vertex can change if restarted. For example, 
1. a restarted source may return all the previous source splits and consume different splits.
2. if a result is re-produced differently, its restarted consumers may produce different results.

Besides that, I think there's a problem with fix for my above question `#2`. Because `taskExecutionState.getExecutionState() == ExecutionState.FINISHED` does not mean the task truly transitions to `FINISHED` at JM side.

Reply:
> Looks to me it's not right because the data produced by an execution vertex can change if 

You are right. It looks like we have to record the subpartition bytes for each subpartition (no aggregation), then we can reset the bytes when a partition is reset. 
However, once all partitions are finished, we can convert the subpartiton bytes to aggregated value(to reduce the space usage). Because the distribution of source splits does not affect the distribution of data consumed by downstream tasks of ALL_TO_ALL edges(Hashing or Rebalancing, we do not consider rare cases such as custom partitions here). 

> Because `taskExecutionState.getExecutionState() == ExecutionState.FINISHED` does not mean the task truly transitions to FINISHED at JM side.

Yes, we should check whether the state update was successful.